File: app/src/main/java/com/example/firefighters/repositories/UserRepository.java
package com.example.firefighters.repositories;

import android.app.Activity;

import com.example.firefighters.models.EmergencyModel;
import com.example.firefighters.models.UserModel;
import com.example.firefighters.tools.ConstantsValues;
import com.example.firefighters.tools.FirebaseManager;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.EventListener;
import com.google.firebase.firestore.FirebaseFirestoreException;
import com.google.firebase.firestore.QuerySnapshot;

import org.jetbrains.annotations.NotNull;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

public class UserRepository {

    private static UserRepository instance;

    public static UserRepository getInstance() {
        if (instance == null)
            instance = new UserRepository();
        return instance;
    }

    //Sign in with mail and password
    public LiveData<Integer> signInUser(String userMail, String userPassword) {
        MutableLiveData<Integer> data = new MutableLiveData<>();
        if (userMail == null || userPassword == null || userMail.isEmpty() || userPassword.isEmpty()) {
            data.setValue(-1);
        } else {
            FirebaseManager.getInstance().getFirebaseAuthInstance().signInWithEmailAndPassword(userMail, userPassword)
                    .addOnCompleteListener(new OnCompleteListener<AuthResult>() {
                        @Override
                        public void onComplete(@NonNull @NotNull Task<AuthResult> task) {
                            if (task.isSuccessful()) {
                                data.setValue(1);
                            } else {
                                data.setValue(-1);
                            }
                        }
                    });
        }
        return data;
    }

    //Create new user
    public LiveData<Integer> createNewUser(String userName, String userMail, String userPassword, boolean isFirefighter) {
        MutableLiveData<Integer> data = new MutableLiveData<>();
        if (userMail == null || userPassword == null || userMail.isEmpty() || userPassword.isEmpty()) {
            data.setValue(-1);
        } else {
            FirebaseManager.getInstance().getFirebaseAuthInstance().createUserWithEmailAndPassword(userMail, userPassword)
                    .addOnCompleteListener(new OnCompleteListener<AuthResult>() {
                        @Override
                        public void onComplete(@NonNull @NotNull Task<AuthResult> task) {
                            if (task.isSuccessful()) {
                                if (userMail.equals(ConstantsValues.ADMIN_EMAIL)) {
                                    data.setValue(2);
                                } else {
                                    FirebaseManager.getInstance().getFirebaseFirestoreInstance()
                                            .collection(ConstantsValues.USERS_COLLECTION)
                                            .whereEqualTo("mail", userMail)
                                            .get()
                                            .addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
                                                @Override
                                                public void onComplete(@NonNull @NotNull Task<QuerySnapshot> task) {
                                                    if (task.isSuccessful()) {
                                                        if (task.getResult().size() > 0) {
                                                            UserModel user = task.getResult().getDocuments().get(0).toObject(UserModel.class);
                                                            if(user != null && user.isFireFighter()){
                                                                data.setValue(1);
                                                            }else if (user != null){
                                                                data.setValue(-1);
                                                            }
                                                        } else {
                                                            UserModel currentUserModel = new UserModel();
                                                            currentUserModel.setUserName(userName);
                                                            currentUserModel.setFireFighter(isFirefighter);
                                                            currentUserModel.setMail(userMail);
                                                            FirebaseManager.getInstance().getFirebaseFirestoreInstance()
                                                                    .collection(ConstantsValues.USERS_COLLECTION)
                                                                    .add(currentUserModel)
                                                                    .addOnCompleteListener(new OnCompleteListener<DocumentReference>() {
                                                                        @Override
                                                                        public void onComplete(@NonNull @NotNull Task<DocumentReference> task) {
                                                                            if (task.isSuccessful()) {
                                                                                data.setValue(1);
                                                                            } else {
                                                                                data.setValue(-1);
                                                                            }
                                                                        }
                                                                    });
                                                        }
                                                    } else {
                                                        data.setValue(-1);
                                                    }
                                                }
                                            });
                                }
                            } else {
                                data.setValue(-1);
                            }
                        }
                    });
        }
        return data;
    }

    public LiveData<UserModel> loadUserModel(String userMail) {
        MutableLiveData<UserModel> data = new MutableLiveData<>();
        UserModel user = null;
        if (userMail == null || userMail.isEmpty()) {
            data.setValue(null);
        } else {
            FirebaseManager.getInstance().getFirebaseFirestoreInstance()
                    .collection(ConstantsValues.USERS_COLLECTION)
                    .whereEqualTo("mail", userMail)
                    .get()
                    .addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
                        @Override
                        public void onComplete(@NonNull @NotNull Task<QuerySnapshot> task) {
                            if (task.isSuccessful()) {
                                if (task.getResult().size() > 0) {
                                    UserModel user = task.getResult().getDocuments().get(0).toObject(UserModel.class);
                                    data.setValue(user);
                                } else {
                                    data.setValue(null);
                                }
                            } else {
                                data.setValue(null);
                            }
                        }
                    });
        }
        return data;
    }


    public LiveData<QuerySnapshot> loadAllFireFightersWorkingUnitQuery(String unit) {
        MutableLiveData<QuerySnapshot> data = new MutableLiveData<>();
        FirebaseManager.getInstance().getFirebaseFirestoreInstance()
                .collection(ConstantsValues.USERS_COLLECTION)
                .whereEqualTo("unit", unit)
                .get()
                .addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
                    @Override
                    public void onComplete(@NonNull @NotNull Task<QuerySnapshot> task) {
                        data.setValue(task.getResult());
                    }
                });
        return data;
    }

    //Reset password
    public LiveData<Integer> resetPasswordWithMail(Activity activity, String userMail) {
        MutableLiveData<Integer> data = new MutableLiveData<>();
        FirebaseManager.getInstance().getFirebaseAuthInstance().sendPasswordResetEmail(userMail)
                .addOnCompleteListener(new OnCompleteListener<Void>() {
                    @Override
                    public void onComplete(@NonNull @NotNull Task<Void> task) {
                        if (task.isSuccessful()) {
                            data.setValue(1);
                        } else {
                            data.setValue(-1);
                        }
                    }
                });
        return data;
    }

    public LiveData<Integer> logOut() {
        MutableLiveData<Integer> data = new MutableLiveData<>();
        FirebaseManager.getInstance().getFirebaseAuthInstance().signOut();
        data.setValue(1);
        return data;
    }

    public LiveData<QuerySnapshot> loadFireFightersQuery(DocumentSnapshot lastDocument, int limitCount) {
        MutableLiveData<QuerySnapshot> data = new MutableLiveData<>();
        if (lastDocument != null){
            FirebaseManager.getInstance().getFirebaseFirestoreInstance()
                    .collection(ConstantsValues.USERS_COLLECTION)
                    .whereEqualTo("fireFighter", true)
                    .startAfter(lastDocument)
                    .limit(limitCount)
                    .get()
                    .addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
                        @Override
                        public void onComplete(@NonNull @NotNull Task<QuerySnapshot> task) {
                            data.setValue(task.getResult());
                        }
                    });
        }else{
            FirebaseManager.getInstance().getFirebaseFirestoreInstance()
                    .collection(ConstantsValues.USERS_COLLECTION)
                    .whereEqualTo("fireFighter", true)
                    .limit(limitCount)
                    .get()
                    .addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
                        @Override
                        public void onComplete(@NonNull @NotNull Task<QuerySnapshot> task) {
                            data.setValue(task.getResult());
                        }
                    });
        }
        return data;
    }

    public LiveData<QuerySnapshot> loadFireFightersQuerySnapshot() {
        MutableLiveData<QuerySnapshot> data = new MutableLiveData<>();
        FirebaseManager.getInstance().getFirebaseFirestoreInstance()
                .collection(ConstantsValues.USERS_COLLECTION)
                .whereEqualTo("fireFighter", true)
                .addSnapshotListener(new EventListener<QuerySnapshot>() {
                    @Override
                    public void onEvent(@Nullable @org.jetbrains.annotations.Nullable QuerySnapshot value, @Nullable @org.jetbrains.annotations.Nullable FirebaseFirestoreException error) {
                        data.setValue(value);
                    }
                });
        return data;
    }

    public LiveData<Integer> saveFireFighter(UserModel firefighterModel) {
        MutableLiveData<Integer> data = new MutableLiveData<>();
        FirebaseManager.getInstance().getFirebaseFirestoreInstance()
                .collection(ConstantsValues.USERS_COLLECTION)
                .add(firefighterModel)
                .addOnCompleteListener(new OnCompleteListener<DocumentReference>() {
                    @Override
                    public void onComplete(@NonNull @NotNull Task<DocumentReference> task) {
                        if (task.isSuccessful()){
                            data.setValue(1);
                        }else{
                            data.setValue(-1);
                        }
                    }
                });
        return data;
    }

    public LiveData<Integer> updateFireFighter(UserModel firefighterModel) {
        MutableLiveData<Integer> data = new MutableLiveData<>();
        FirebaseManager.getInstance().getFirebaseFirestoreInstance()
                .collection(ConstantsValues.EMERGENCIES_COLLECTION)
                .whereEqualTo("mail", firefighterModel.getMail())
                .get()
                .addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
                    @Override
                    public void onComplete(@NonNull @NotNull Task<QuerySnapshot> task) {
                        if (task.isSuccessful()){
                            data.setValue(1);
                        }else{
                            data.setValue(-1);
                        }
                    }
                });
        return data;
    }

    public LiveData<Integer> deleteFireFighter(UserModel firefighterModel) {
        MutableLiveData<Integer> data = new MutableLiveData<>();
        FirebaseManager.getInstance().getFirebaseFirestoreInstance()
                .collection(ConstantsValues.USERS_COLLECTION)
                .whereEqualTo("fireFighter", true)
                .whereEqualTo("mail", firefighterModel.getMail())
                .get()
                .addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
                    @Override
                    public void onComplete(@NonNull @NotNull Task<QuerySnapshot> task) {
                        if (task.isSuccessful()){
                            for (DocumentSnapshot document: task.getResult()) {
                                document.getReference()
                                        .delete()
                                        .addOnCompleteListener(new OnCompleteListener<Void>() {
                                            @Override
                                            public void onComplete(@NonNull @NotNull Task<Void> task) {
                                                if(task.isSuccessful()){
                                                    data.setValue(1);
                                                }else{
                                                    data.setValue(-1);
                                                }
                                            }
                                        });
                            }
                        }else{
                            data.setValue(-1);
                        }
                    }
                });
        return data;
    }
}

File: app/src/main/java/com/example/firefighters/repositories/UserRepository.kt
package com.example.firefighters.repositories

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.example.firefighters.models.UserModel
import com.example.firefighters.utils.ConstantsValues
import com.example.firefighters.utils.FirebaseUtils
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.QuerySnapshot

class UserRepository {
    //Sign in with mail and password
    fun signInUser(userMail: String?, userPassword: String?): LiveData<Int> {
        val data = MutableLiveData<Int>()
        if (userMail == null || userPassword == null || userMail.isEmpty() || userPassword.isEmpty()) {
            data.setValue(-1)
        } else {
            FirebaseUtils.instance?.firebaseAuthInstance
                ?.signInWithEmailAndPassword(
                userMail,
                userPassword
            )
                ?.addOnCompleteListener { task ->
                    if (task.isSuccessful) {
                        data.setValue(1)
                    } else {
                        data.setValue(-1)
                    }
                }
        }
        return data
    }

    //Create new user
    fun createNewUser(
        userName: String?,
        userMail: String?,
        userPassword: String?,
        isFirefighter: Boolean
    ): LiveData<Int> {
        val data = MutableLiveData<Int>()
        if (userMail == null || userPassword == null || userMail.isEmpty() || userPassword.isEmpty()) {
            data.setValue(-1)
        } else {
            FirebaseUtils.instance?.firebaseAuthInstance
                ?.createUserWithEmailAndPassword(
                userMail,
                userPassword
            )
                ?.addOnCompleteListener { task ->
                    if (task.isSuccessful) {
                        if (userMail == ConstantsValues.ADMIN_EMAIL) {
                            data.setValue(2)
                        } else {
                            FirebaseUtils.instance?.firebaseFirestoreInstance
                                ?.collection(ConstantsValues.USERS_COLLECTION)
                                ?.whereEqualTo("mail", userMail)
                                ?.get()
                                ?.addOnCompleteListener {
                                    if (it.isSuccessful) {
                                        if ((it.result?.size() ?: 0) > 0) {
                                            val user = it.result?.documents?.get(0)?.toObject(
                                                UserModel::class.java
                                            )
                                            if (user != null && user.isFireFighter) {
                                                data.setValue(1)
                                            } else if (user != null) {
                                                data.value = -1
                                            }
                                        } else {
                                            val currentUserModel = UserModel()
                                            currentUserModel.userName = userName
                                            currentUserModel.isFireFighter = isFirefighter
                                            currentUserModel.mail = userMail
                                            FirebaseUtils.instance?.firebaseFirestoreInstance
                                                ?.collection(ConstantsValues.USERS_COLLECTION)
                                                ?.add(currentUserModel)
                                                ?.addOnCompleteListener {
                                                    if (it.isSuccessful) {
                                                        data.setValue(1)
                                                    } else {
                                                        data.setValue(-1)
                                                    }
                                                }
                                        }
                                    } else {
                                        data.setValue(-1)
                                    }
                                }
                        }
                    } else {
                        data.setValue(-1)
                    }
                }
        }
        return data
    }

    fun loadUserModel(userMail: String?): LiveData<UserModel?> {
        val data = MutableLiveData<UserModel?>()
        if (userMail.isNullOrEmpty()) {
            data.setValue(null)
        } else {
            FirebaseUtils.instance?.firebaseFirestoreInstance
                ?.collection(ConstantsValues.USERS_COLLECTION)
                ?.whereEqualTo("mail", userMail)
                ?.get()
                ?.addOnCompleteListener { task ->
                    if (task.isSuccessful) {
                        if ((task.result?.size() ?: 0) > 0) {
                            data.setValue(
                                task.result?.documents?.get(0)?.toObject(
                                    UserModel::class.java
                                )
                            )
                        } else {
                            data.setValue(null)
                        }
                    } else {
                        data.setValue(null)
                    }
                }
        }
        return data
    }

    fun loadAllFireFightersWorkingUnitQuery(unit: String?): LiveData<QuerySnapshot> {
        val data = MutableLiveData<QuerySnapshot>()
        FirebaseUtils.instance?.firebaseFirestoreInstance
            ?.collection(ConstantsValues.USERS_COLLECTION)
            ?.whereEqualTo("unit", unit)
            ?.get()
            ?.addOnCompleteListener { task -> data.value = task.result }
        return data
    }

    //Reset password
    fun resetPasswordWithMail(userMail: String?): LiveData<Int> {
        val data = MutableLiveData<Int>()
        FirebaseUtils.instance?.firebaseAuthInstance
            ?.sendPasswordResetEmail(userMail!!)
            ?.addOnCompleteListener { task ->
                if (task.isSuccessful) {
                    data.setValue(1)
                } else {
                    data.setValue(-1)
                }
            }
        return data
    }

    fun logOut(): LiveData<Int> {
        val data = MutableLiveData<Int>()
        FirebaseUtils.instance?.firebaseAuthInstance
            ?.signOut()
        data.value = 1
        return data
    }

    fun loadFireFightersQuery(
        lastDocument: DocumentSnapshot?,
        limitCount: Int
    ): LiveData<QuerySnapshot> {
        val data = MutableLiveData<QuerySnapshot>()
        if (lastDocument != null) {
            FirebaseUtils.instance?.firebaseFirestoreInstance
                ?.collection(ConstantsValues.USERS_COLLECTION)
                ?.whereEqualTo("fireFighter", true)
                ?.startAfter(lastDocument)
                ?.limit(limitCount.toLong())
                ?.get()
                ?.addOnCompleteListener { task -> data.value = task.result }
        } else {
            FirebaseUtils.instance?.firebaseFirestoreInstance
                ?.collection(ConstantsValues.USERS_COLLECTION)
                ?.whereEqualTo("fireFighter", true)
                ?.limit(limitCount.toLong())
                ?.get()
                ?.addOnCompleteListener { task -> data.value = task.result }
        }
        return data
    }

    fun loadFireFightersQuerySnapshot(): LiveData<QuerySnapshot?> {
        val data = MutableLiveData<QuerySnapshot?>()
        FirebaseUtils.instance?.firebaseFirestoreInstance
            ?.collection(ConstantsValues.USERS_COLLECTION)
            ?.whereEqualTo("fireFighter", true)
            ?.addSnapshotListener { value, _ -> data.value = value }
        return data
    }

    fun saveFireFighter(firefighterModel: UserModel?): LiveData<Int> {
        val data = MutableLiveData<Int>()
        FirebaseUtils.instance?.firebaseFirestoreInstance
            ?.collection(ConstantsValues.USERS_COLLECTION)
            ?.add(firefighterModel!!)
            ?.addOnCompleteListener { task ->
                if (task.isSuccessful) {
                    data.setValue(1)
                } else {
                    data.setValue(-1)
                }
            }
        return data
    }

    fun updateFireFighter(firefighterModel: UserModel): LiveData<Int> {
        val data = MutableLiveData<Int>()
        FirebaseUtils.instance?.firebaseFirestoreInstance
            ?.collection(ConstantsValues.EMERGENCIES_COLLECTION)
            ?.whereEqualTo("mail", firefighterModel.mail)
            ?.get()
            ?.addOnCompleteListener { task ->
                if (task.isSuccessful) {
                    data.setValue(1)
                } else {
                    data.setValue(-1)
                }
            }
        return data
    }

    fun deleteFireFighter(firefighterModel: UserModel): LiveData<Int> {
        val data = MutableLiveData<Int>()
        FirebaseUtils.instance?.firebaseFirestoreInstance
            ?.collection(ConstantsValues.USERS_COLLECTION)
            ?.whereEqualTo("fireFighter", true)
            ?.whereEqualTo("mail", firefighterModel.mail)
            ?.get()
            ?.addOnCompleteListener { task ->
                if (task.isSuccessful) {
                    for (document in task.result ?: listOf()) {
                        document.reference
                            .delete()
                            .addOnCompleteListener {
                                if (it.isSuccessful) {
                                    data.setValue(1)
                                } else {
                                    data.setValue(-1)
                                }
                            }
                    }
                } else {
                    data.setValue(-1)
                }
            }
        return data
    }
}
